File: src/main/java/me/hhitt/disasters/disaster/impl/BlockDisappear.java
package me.hhitt.disasters.disaster.impl;

import me.hhitt.disasters.arena.Arena;
import me.hhitt.disasters.disaster.Disaster;
import me.hhitt.disasters.disaster.DisasterRegistry;
import me.hhitt.disasters.model.block.DisappearBlock;
import me.hhitt.disasters.util.Notify;
import org.bukkit.Location;
import org.bukkit.entity.Player;

import java.util.concurrent.ConcurrentHashMap;

public class BlockDisappear implements Disaster {

    private final ConcurrentHashMap<String, DisappearBlock> blocks = new ConcurrentHashMap<>();
    private Arena arena;
    private final int duration = 120;
    private int elapsed = 0;
    private boolean active = true;

    @Override
    public void start(Arena arena) {
        this.arena = arena;
        Notify.disaster(arena, "disappear-blocks");
    }

    @Override
    public void pulse(int time) {
        elapsed++;
        if (elapsed >= duration) {
            active = false;
        }
        for (DisappearBlock block : blocks.values()) {
            block.updateMaterial();
        }

        Arena currentArena = arena;
        if (currentArena == null) return;
        if (!active) return;
        for (Player player : currentArena.getAlive()) {
            Location loc = player.getLocation();
            Location blockBelow = loc.clone().subtract(0.0, 1.0, 0.0);
            if (blockBelow.getBlock().getType().isAir()) {
                double[] offsets = new double[]{-0.3, 0.3};
                outer:
                for (double dx : offsets) {
                    for (double dz : offsets) {
                        Location check = loc.clone().add(dx, -1.0, dz);
                        if (!check.getBlock().getType().isAir() && check.getBlock().getType().isSolid()) {
                            DisasterRegistry.addBlockToDisappear(currentArena, loc.clone().add(dx, 0.0, dz));
                            break outer;
                        }
                    }
                }
            }
        }
    }

    @Override
    public void stop(Arena arena) {
        blocks.clear();
        this.arena = null;
    }

    public boolean isActive() {
        return active;
    }

    public void addBlock(Arena arena, Location location) {
        if (!active) return;
        String key = locationKey(location);
        DisappearBlock existing = blocks.get(key);
        if (existing != null) {
            existing.setOccupied(true);
            return;
        }
        DisappearBlock block = new DisappearBlock(arena, location);
        blocks.put(key, block);
    }

    public void setUnoccupied(Location location) {
        String key = locationKey(location);
        DisappearBlock block = blocks.get(key);
        if (block != null) {
            block.setOccupied(false);
        }
    }

    public void removeBlock(DisappearBlock block) {
        blocks.remove(locationKey(block.getLocation()));
    }

    private String locationKey(Location loc) {
        return loc.getBlockX() + "," + loc.getBlockY() + "," + loc.getBlockZ();
    }
}

File: src/main/kotlin/me/hhitt/disasters/disaster/impl/BlockDisappear.kt
package me.hhitt.disasters.disaster.impl

import me.hhitt.disasters.arena.Arena
import me.hhitt.disasters.disaster.Disaster
import me.hhitt.disasters.disaster.DisasterRegistry
import me.hhitt.disasters.model.block.DisappearBlock
import me.hhitt.disasters.util.Notify
import java.util.concurrent.ConcurrentHashMap

class BlockDisappear : Disaster {

    // Keyed by block location string so we can look up existing blocks for pause/resume
    private val blocks = ConcurrentHashMap<String, DisappearBlock>()
    private var arena: Arena? = null

    // Duration in seconds — disaster stops accepting new blocks after this
    private val duration = 120
    private var elapsed = 0
    private var active = true

    override fun start(arena: Arena) {
        this.arena = arena
        Notify.disaster(arena, "disappear-blocks")
    }

    override fun pulse(time: Int) {
        elapsed++
        // Stop accepting new blocks after duration, but let existing ones finish breaking
        if (elapsed >= duration) {
            active = false
        }
        blocks.values.forEach { it.updateMaterial() }

        // If a player is standing on air, find the closest solid block under them and add it
        val currentArena = arena ?: return
        if (!active) return
        currentArena.alive.toList().forEach { player ->
            val loc = player.location
            val blockBelow = loc.clone().subtract(0.0, 1.0, 0.0)
            if (blockBelow.block.type.isAir) {
                // Player center is over air — check hitbox corners for a solid block nearby
                val offsets = doubleArrayOf(-0.3, 0.3)
                for (dx in offsets) {
                    for (dz in offsets) {
                        val check = loc.clone().add(dx, -1.0, dz)
                        if (!check.block.type.isAir && check.block.type.isSolid) {
                            DisasterRegistry.addBlockToDisappear(currentArena, loc.clone().add(dx, 0.0, dz))
                            return@forEach
                        }
                    }
                }
            }
        }
    }

    override fun stop(arena: Arena) {
        blocks.clear()
        this.arena = null
    }

    fun isActive(): Boolean = active

    /**
     * Adds a block or marks an existing one as occupied (resumes cracking).
     * Uses location key to prevent duplicate entries for the same block.
     */
    fun addBlock(arena: Arena, location: org.bukkit.Location) {
        if (!active) return
        val key = locationKey(location)
        val existing = blocks[key]
        if (existing != null) {
            // Player stepped back on a block that was already cracking — resume
            existing.occupied = true
            return
        }
        val block = DisappearBlock(arena, location)
        blocks[key] = block
    }

    /**
     * Marks a block as unoccupied (pauses breaking) when the player steps off.
     */
    fun setUnoccupied(location: org.bukkit.Location) {
        val key = locationKey(location)
        blocks[key]?.occupied = false
    }

    fun removeBlock(block: DisappearBlock) {
        blocks.remove(locationKey(block.location))
    }

    private fun locationKey(loc: org.bukkit.Location): String {
        return "${loc.blockX},${loc.blockY},${loc.blockZ}"
    }
}